File: app/src/main/java/com/ne1c/gitteroid/api/GitterStreamer.java
package com.ne1c.gitteroid.api;

import com.google.gson.Gson;
import com.ne1c.gitteroid.models.data.MessageModel;
import com.ne1c.gitteroid.utils.Utils;

import java.io.BufferedReader;
import java.io.IOException;

import javax.inject.Inject;

import rx.Observable;
import rx.Observer;
import rx.observables.SyncOnSubscribe;

public class GitterStreamer {
    private final String STREAM_URL = "https://stream.gitter.im/v1/rooms/%s/chatMessages";

    private GitterStreamApi mStreamApi;

    @Inject
    public GitterStreamer(GitterStreamApi api) {
        mStreamApi = api;
    }

    public Observable<MessageModel> getMessageStream(String roomId) {
        String streamUrl = String.format(STREAM_URL, roomId);
        return mStreamApi.getMessagesStream(streamUrl, Utils.getInstance().getBearer())
                .flatMap(response -> Observable.create(new OnSubscribeBufferedReader(new BufferedReader(response.charStream())))).filter(s -> s != null && !s.trim().isEmpty())
                .map(s -> new Gson().fromJson(s, MessageModel.class));
    }

    public class OnSubscribeBufferedReader extends SyncOnSubscribe<BufferedReader, String> {
        private final BufferedReader mReader;

        public OnSubscribeBufferedReader(BufferedReader reader) {
            mReader = reader;
        }


        @Override
        protected BufferedReader generateState() {
            return mReader;
        }

        @Override
        protected BufferedReader next(BufferedReader state, Observer<? super String> observer) {
            String line;
            try {
                line = state.readLine();
                if (line == null) {
                    observer.onCompleted();
                } else {
                    observer.onNext(line);
                }
            } catch (IOException e) {
                try {
                    state.close();
                } catch (IOException e1) {
                    e1.printStackTrace();
                }
                observer.onError(e);
            }

            return state;
        }
    }
}

File: app/src/main/kotlin/com/ne1c/gitteroid/api/GitterStreamer.kt
package com.ne1c.gitteroid.api

import com.google.gson.Gson
import com.ne1c.gitteroid.models.data.MessageModel
import rx.Observable
import rx.Observer
import rx.observables.SyncOnSubscribe
import java.io.BufferedReader
import java.io.IOException

class GitterStreamer(private val mStreamApi: GitterStreamApi,
                     private val bearer: String) {
    private val STREAM_URL = "https://stream.gitter.im/v1/rooms/%s/chatMessages"

    fun getMessageStream(roomId: String): Observable<MessageModel> {
        val streamUrl = String.format(STREAM_URL, roomId)
        return mStreamApi.getMessagesStream(streamUrl, bearer)
                .flatMap({ Observable.create(OnSubscribeBufferedReader(BufferedReader(it.charStream()))) })
                .filter { it != null && !it.trim { it <= ' ' }.isEmpty() }
                .map({ Gson().fromJson<MessageModel>(it, MessageModel::class.java) })
    }

    private inner class OnSubscribeBufferedReader(private val mReader: BufferedReader) : SyncOnSubscribe<BufferedReader, String>() {
        override fun generateState(): BufferedReader {
            return mReader
        }

        override fun next(state: BufferedReader, observer: Observer<in String>): BufferedReader {
            val line: String?
            try {
                line = state.readLine()
                if (line == null) {
                    observer.onCompleted()
                } else {
                    observer.onNext(line)
                }
            } catch (e: IOException) {
                try {
                    state.close()
                } catch (e1: IOException) {
                    e1.printStackTrace()
                }

                observer.onError(e)
            }

            return state
        }
    }
}
